webapi/Controllers/ChatHistoryController.cs

```diff
@@ -260,4 +268,103 @@ public async Task<IActionResult> GetAllChatSessionsAsync(string userId)
+        {
+            return this.NotFound($"No chat session found for chat id '{chatId}'.");
+        }
+
+        // Delete any resources associated with the chat session.
+        var deleteResourcesResult = await this.DeleteChatResourcesAsync(sessionId, cancellationToken) as StatusCodeResult;
+        if (deleteResourcesResult?.StatusCode != 204)
+        {
+            return this.StatusCode(500, $"Failed to delete resources for chat id '{chatId}'.");
+        }
+
+        // Delete chat session and broadcast update to all participants.
+        await this._sessionRepository.DeleteAsync(chatToDelete);
+        await messageRelayHubContext.Clients.Group(chatId).SendAsync(ChatDeletedClientCall, chatId, authInfo.UserId, cancellationToken: cancellationToken);
+
+        return this.NoContent();
+    }
+
+    /// <summary>
+    /// Deletes all associated resources (messages, memories, participants) associated with a chat session.
+    /// </summary>
+    /// <param name="sessionId">The chat id.</param>
+    private async Task<IActionResult> DeleteChatResourcesAsync(Guid sessionId, CancellationToken cancellationToken)
+    {
+        var chatId = sessionId.ToString();
+        var cleanupTasks = new List<Task>();
+
+        // Create and store the tasks for deleting all users tied to the chat.
+        var participants = await this._participantRepository.FindByChatIdAsync(chatId);
+        foreach (var participant in participants)
+        {
+            cleanupTasks.Add(this._participantRepository.DeleteAsync(participant));
+        }
+
+        // Create and store the tasks for deleting chat messages.
+        var messages = await this._messageRepository.FindByChatIdAsync(chatId);
+        foreach (var message in messages)
+        {
+            cleanupTasks.Add(this._messageRepository.DeleteAsync(message));
+        }
+
+        // Create and store the tasks for deleting memory sources.
+        var sources = await this._sourceRepository.FindByChatIdAsync(chatId, false);
+        foreach (var source in sources)
+        {
+            cleanupTasks.Add(this._sourceRepository.DeleteAsync(source));
+        }
+
+        // Create and store the tasks for deleting semantic memories.
+        // TODO: [Issue #47] Filtering memory collections by name might be fragile.
+        var memoryCollections = (await this._memoryStore.GetCollectionsAsync(cancellationToken).ToListAsync<string>())
+            .Where(collection => collection.StartsWith(chatId, StringComparison.OrdinalIgnoreCase));
+        foreach (var collection in memoryCollections)
+        {
+            cleanupTasks.Add(this._memoryStore.DeleteCollectionAsync(collection, cancellationToken));
+        }
+
+        // Await all the tasks in parallel and handle the exceptions
+        await Task.WhenAll(cleanupTasks);
+        var failedTasks = false;
+
+        // Iterate over the tasks and check their status and exception
+        foreach (var task in cleanupTasks)
```

Comment:
If any of your tasks raises an exception, you will not reach this code.

See following for details:
https://www.davidguida.net/exception-handling-in-parallel-tasks/

Reply:
Changed to handle exceptions in a try/catch block